### DISCUSSION TITLE
Autocomplete FluentValidations

### DISCUSSION THREAD

**Discussion by mdmontesinos** [Megabit/Blazorise#5129] [Q&A]
As far as I've read in related issues, currently neither DataAnnotations nor FluentValidations is supported in Autocomplete component.

#4628, #4767, #3537, #4407

Which is the most appropriate/viable way to use FluentValidations making use of the Validator in the Autocomplete?

I was wondering if it's possible to access the FluentValidationHandler defined in the <Validations> component to manually validate the property in the Validator.

Something like:

```razor
<Validations @ref="fluentValidations" Model="@address" Mode="ValidationMode.Auto" ValidateOnLoad="false" HandlerType="typeof(FluentValidationHandler)">
<Validation>
    <Field>
        <FieldLabel RequiredIndicator>País</FieldLabel>
        <FieldBody>
            <Autocomplete Class="autocomplete"
                          TItem="string"
                          TValue="string"
                          Data="@Countries"
                          TextField="@(( item ) => item)"
                          ValueField="@(( item ) => item)"
                          Placeholder="País"
                          CloseOnSelection="false"
                          MinLength="0"
                          FreeTyping="false"
                          Validator="TestAutocompleteValidator"
            @bind-SelectedValue="address.Country">
                <NotFoundContent>No se han encontrado países</NotFoundContent>
            </Autocomplete>
        </FieldBody>
    </Field>
</Validation>
</Validations>

@code {

    Address address= new();

    Validations? fluentValidations;

    private void TestAutocompleteValidator(ValidatorEventArgs e)
    {
        //Something like this
        e.Status = fluentValidations.Validate(address.Country);
    }

    List<string> Countries = new()
    {
        "España", "Portugal", "Francia", "Alemania"
    };
}
```

**Comment by stsrki** [comment]
The problem with that approach is that Autocomplete already has a Validation component defined internally. So it will not be able to override it.

I will create a ticket from this discussion and investigate what can be done. Maybe we can just add new HandlerType parameter and then propagate it to the internal Validation.

**Reply by mdmontesinos** [reply]
Thanks! 

I feel this issue is somehow "urgent" and it's been lingering for some time now, as DataAnnotations and Regular Validator are rendered mostly useless for complex business requirements, and Autocomplete is the ideal component in forms where you need to select from large lists (countries, states, cities, ...)

**Reply by David-Moreira** [reply]
@stsrki I'd personally consider this:

https://github.com/Megabit/Blazorise/issues/4767

![image](https://github.com/Megabit/Blazorise/assets/22283161/cbc4a7e4-faa3-43b1-af0b-adb5054d1b4b)

It would be more inline with how Blazorise works in general, which would be `<validation><component/></validation>` allowing as much flexibility as any other component has by having the `Validation` component defined "outside".

By having it defined internally, we'll always be playing "catch up" by having to introduce new parameters, etc..

**Reply by stsrki** [reply]
The problem with that approach is that still validates on internal TextEdit component that is used to search the values. Also Autocomplete internally uses a Dropdown which doesn't have any validation support.

I think we would have to add fully validation to the Autocomplete so that it behaves more like a Select component, validation and selection wise. In any case, it would a breaking change if we would to remove an internal Validation.

**Reply by David-Moreira** [reply]
> The problem with that approach is that still validates on internal TextEdit component that is used to search the values. Also > Autocomplete internally uses a Dropdown which doesn't have any validation support.

> I think we would have to add fully validation to the Autocomplete so that it behaves more like a Select component, validation and selection wise. In any case, it would a breaking change if we would to remove an internal Validation.

What's the problem exactly? Yes I would take the breaking change. I think it makes sense in order to make it behave like the other components making the library consistent. Currently, to me it is not consistent.
Of course, we'd need to do it in v2.0

**Reply by stsrki** [reply]
If you look into the Autocomplete code, you will see that we wrap the TextEdit.

![image](https://github.com/Megabit/Blazorise/assets/900302/ab254782-9ed8-4ebc-bc16-7d4ef4fb6836)

Only the TextEdit is validated and it is validated only with the `Validator` and `AsyncValidator`. To be able to validate data-annotation we would have to inherit Autocomplete from `IValidationInput`. By inheriting from `IValidationInput` we would also have to implement all the wirings around the `Validation` component. These include `SelectedValue`, `SelectedValueChanged`, and `SelectedValueExpression` (we don't have this at the moment). Also we need to think should we also validate `SelectedText`, and `SelectedTextChanged`. Does it makes sense to validate it? There is also `SelectedValues` and `SelectedValuesChanged`. So all in all there is much stuff that we need to keep in mind.

**Reply by David-Moreira** [reply]
Should validate similar to `Select` in my opinion. As the idea is the same, there's a list of values based on key / values. And you either validate the key or value. I imagine that in `Select` we are validating the key, which is what makes most sense I guess?

But I understand it might be complex to implement, that does not mean we should just give up :D We should tackle all those questions and do it in my opinion.

**Reply by stsrki** [reply]
> But I understand it might be complex to implement, that does not mean we should just give up :D We should tackle all those questions and do it in my opinion.

Agree 100%.
